### PR TITLE
Python harmonic

### PR DESCRIPTION
**Pull request by auggiemarignier**
Carrying on from #22.  Python wrappers for the wavelet transforms and adjoints in harmonic space.  So far only had wrappers for the pixel space transforms.  This goes some way towards #20

### CODE REVIEW

File: src/test/python/test_directional_harmonic.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    analysis_lm2lmn,
+    analysis_adjoint_lmn2lm,
+    synthesis_lmn2lm,
+    synthesis_adjoint_lm2lmn
+)
+
+
+@fixture
+def rng(request):
+    return np.random.default_rng(getattr(request.config.option, "randomly_seed", None))
```
Comment (paddyroddy) [inline src/test/python/test_directional_harmonic.py:17]: Yay for new `numpy.random` 😄

File: src/test/python/test_directional_harmonic.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        (
+            partial(synthesis_adjoint_lm2lmn, spin=0, upsample=1, N=1),
+            partial(synthesis_lmn2lm, spin=0, upsample=1, N=1),
+        ),
+        (
+            partial(synthesis_adjoint_lm2lmn, spin=2, upsample=1, N=1),
+            partial(synthesis_lmn2lm, spin=2, upsample=1, N=1),
+        ),
+    ],
+)
```
Comment (paddyroddy) [inline src/test/python/test_directional_harmonic.py:49]: noice, btw if you're not aware the library [hypothesis](https://hypothesis.readthedocs.io/en/latest/) is great for getting a test to run multiple circumstances

File: src/test/python/test_directional_harmonic.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+def test_axisym_adjoint(
+    px2wav, wav2px, rng: np.random.Generator, L=10, B=2, J_min=2
```
Comment (paddyroddy) [inline src/test/python/test_directional_harmonic.py:51]: `nit`: would be great if typing was used everywhere 😉

File: src/test/python/test_directional_harmonic.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+):
+    from pys2let import pys2let_j_max
```
Comment (paddyroddy) [inline src/test/python/test_directional_harmonic.py]: this should really be at the top

File: src/test/python/test_directional_harmonic.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+    nwvlts = pys2let_j_max(B, L, J_min) - J_min + 1
+
+    x = random_lms(rng, L)
+    y_scal, y_wav = random_wavlet_lms(rng, L, nwvlts)
+
+    y = wav2px(y_wav, y_scal, B, L, J_min)
+    x_wav, x_scal = px2wav(x, B, L, J_min)
+
+    assert y_wav.conj().T @ x_wav + y_scal.conj() @ x_scal == approx(y.conj().T @ x)
```
Comment (paddyroddy) [inline src/test/python/test_directional_harmonic.py]: could this be made more readable?

**Review by paddyroddy — CHANGES_REQUESTED** [review]
a few comments, otherwise good

File: src/test/python/test_directional_harmonic.py (hunk 4 above)
Comment (auggiemarignier) [reply src/test/python/test_directional_harmonic.py]: `1a81b64`

File: src/test/python/test_directional_harmonic.py (hunk 5 above)
Comment (auggiemarignier) [reply src/test/python/test_directional_harmonic.py]: `b185675`

**Review by paddyroddy — APPROVED** [review]
👍

**Review by mdavezac — APPROVED** [review]

**Comment by auggiemarignier** [comment]
who has the authority to merge and deploy this? @mdavezac @jasonmcewen

**Pull request merged**